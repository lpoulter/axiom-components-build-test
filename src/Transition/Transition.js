import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import { animations } from '@brandwatch/axiom-materials';
import { Base } from '@brandwatch/axiom-components';
import './Transition.css';

export default class Transition extends Component {
  static propTypes = {
    activeIndex: PropTypes.number.isRequired,
    children: PropTypes.node.isRequired,
    height: PropTypes.string.isRequired,
    width: PropTypes.string.isRequired,
  };

  constructor(props) {
    super(props);
    this.previousIndex = props.activeIndex;
  }

  UNSAFE_componentWillUpdate(nextProps) {
    if (nextProps.activeIndex !== this.props.activeIndex) {
      this.previousIndex = this.props.activeIndex;
    }
  }

  render() {
    const {
      activeIndex,
      children,
      height,
      width,
      ...rest
    } = this.props;

    const directionNext = this.previousIndex < activeIndex;
    const duration = animations.transitionTimeSlow;
    const className = directionNext ? 'ax-animate-next' : 'ax-animate-previous';
    const style = {
      animationDuration: `${duration}ms`,
    };

    // https://medium.com/lalilo/dynamic-transitions-with-react-router-and-react-transition-group-69ab795815c9
    const childFactoryCreator = (classNames) => (
      (child) => (
        React.cloneElement(child, {
          classNames,
        })
      )
    );

    return (
      <TransitionGroup
          childFactory={ childFactoryCreator(className) }
          className="ax-transition__wrapper"
          component={ Base }
          style={ { height, width } }
          { ...rest }>
        <CSSTransition classNames={ className } key={ activeIndex } timeout={ duration }>
          <div className="ax-transition__item" style={ style }>
            { children[activeIndex] }
          </div>
        </CSSTransition>
      </TransitionGroup>
    );
  }
}
