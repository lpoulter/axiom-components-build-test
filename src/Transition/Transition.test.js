import React from 'react';
import renderer from 'react-test-renderer';
import Transition from './Transition';

const getComponent = (props = {}) =>
  renderer.create(
    <Transition height="4rem" width="4rem" { ...props }>
      <div className="first" />
      <div className="second" />
    </Transition>
  );

describe('Transition', () => {
  it('renders first child', () => {
    const component = getComponent({ activeIndex: 0 });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders second child', () => {
    const component = getComponent({ activeIndex: 1 });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
