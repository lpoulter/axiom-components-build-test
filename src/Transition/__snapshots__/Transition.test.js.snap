// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transition renders first child 1`] = `
<div
  className="ax-transition__wrapper"
  style={
    Object {
      "height": "4rem",
      "width": "4rem",
    }
  }
>
  <div
    className="ax-transition__item"
    style={
      Object {
        "animationDuration": "400ms",
      }
    }
  >
    <div
      className="first"
    />
  </div>
</div>
`;

exports[`Transition renders second child 1`] = `
<div
  className="ax-transition__wrapper"
  style={
    Object {
      "height": "4rem",
      "width": "4rem",
    }
  }
>
  <div
    className="ax-transition__item"
    style={
      Object {
        "animationDuration": "400ms",
      }
    }
  >
    <div
      className="second"
    />
  </div>
</div>
`;
